### PR TITLE
[ConstraintSystem] Diagnosing invalid explicit coercion via fix 

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
Adding force down fix to explicit` CoerceExpr` and handle invalid explicit coercions on `MissingExplicitConversionFailure` to diagnose invalid explicit coercions via fix and no longer going fall-through into CSDiag. 

References #26710 

cc @xedin

### CODE REVIEW

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [inline lib/Sema/CSSimplify.cpp]: I don't think that this a good place for this logic, it should be in `repairFailures` where other fixes are located.

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [inline lib/Sema/CSSimplify.cpp]: What is the correct correct fix here? Once this is in `repairFailures` you can use `if (hasRestrictionOrConversion(ConversionRestrictionKind::Existential)) break;` instead because conversion to existential type would record `Existential` restriction.

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [inline lib/Sema/CSSimplify.cpp]: Nit: Formatting

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: So we have to add this fix there? Because for what I could see if we don't add the fix here it actually was going directly in CSDiag not even calling the `repairFailures` which is causing the re-typecheck. 

I'm debugging using the example 
```swift
Double(1) as Double as String
```

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: My understanding of this was: add a fix on match types (but not for existential and deep equality because the correct fixes will be added later), so the diagnose would be handled by that fix(avoiding to fall into CSDiag) and since it would be always emitted, there was nothing to change in repair failures because it would end up there anyway.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Sorry I can't really debug this right now but the idea for this I have is - add a case in `repairFailures` when the last locator element is `ExplicitTypeCoercion` unless it's a `DeepEquality` or `Existential` and none of the types are `is<TypeVariableType>()` record a `ForceDowncast` fix. Audit other places where `FourceDowncast` is used and _maybe_ remove them as obsolete.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Your effort to add a warning about unnecessary casts exposed shortcomings of `ForceDowncast` fix which we need to address before other changes could be merged to avoid regressions.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: @xedin Thank you for all the reviews :))
I'll check again as soon as possible but, as far as got into debugging this , for this specific case, it doesn't call `repairFailures` because in the `applySolution` it was not being diagnosed via fix, making it fall into CSDiag, re-typecheck(each child independently) and only then get into `repairFailures`, but at this point it already did the re-typecheck we are trying to avoid. So because of that I thought the place to add the fix was in match types to handle diagnostic via fix early for ExplicitCoercion. 

Also, the only reference to `ForceDowncast::create` is in `repairViaBridgingCast` that is already called on `repairFailures`.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: The idea is to figure out why it's not being diagnosed via a fix and improve that situation :) From debug output of the constraint system I can see that `Double(1) as Double as String` does produce `ForceDowncast` fix but for some reason it's not diagnosed so the first step is to figure out why is that because fix locator points to `CoerceExpr` directly which is not "complex", so there is something else going on here.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Yeah, I remember seeing yesterday that the diagnostic was actually being emitted via the fix, but only after the CSDiag and re-typecheck each child. (I think that's why it is working that way is implemented right now by only recording the fix early instead of wait for repair failures `repairViaBridgingCast`) 
I have to confirm that, maybe it was for other expression, not sure.

I'll resume the work on this later today after work.
Thank's for taking a look on this @xedin :))

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: No problem and thank you for working on this!

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: @xedin I think I've found the issue, 
[This is the point](https://github.com/apple/swift/blob/9b7ab6bfe7d9eb6c129e5de45911e0941d9326d9/lib/Sema/CSApply.cpp#L7660) where it decides if the expression was going to being diagnosed by fixes.
And those are the fixes recorded at that point 
```
Fixes:
  [fix: as to as!] @ locator@0x1178474b0 [Coerce@..swift/test/expr/cast/sr11295.swift:3:21]
```
This diagnostic fails because of an unresolved checked cast and there explicit tell that 
diagnoseFailureForExpr should handle this, so then ending up on CSDiag. 

So the path I'm following right now is trying to find where this fix is being recorded and conditionally add for this kind of expression a ForceDownCast fix instead. 
Do you think that makes sense?

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Also, if this diagnostic failure [here](https://github.com/apple/swift/blob/9b7ab6bfe7d9eb6c129e5de45911e0941d9326d9/lib/Sema/CSApply.cpp#L7660) defer to `repairFailures` before try to go into `CSDiag`

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: Sorry @LucianoPAlmeida, I have on vacation until today. The location you mentioned https://github.com/apple/swift/blob/9b7ab6bfe7d9eb6c129e5de45911e0941d9326d9/lib/Sema/CSApply.cpp#L7660 doesn't have `diagnosedErrorsViaFixes` because the call to `MissingExplicitConversionFailure` from `ForceDowncast::diagnose` didn't succeed. I'm not sure why is that since locator is not "complex" - it's a single `CoerceExpr`, so what we need to figure out is how to make it so `MissingExplicitConversionFailure` produces a diagnostic in this particular scenario.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: > Sorry @LucianoPAlmeida, I have on vacation until today. The location you mentioned

No problem @xedin, I'm still working on this but not able to make much progress this weekend because of [build issues](https://forums.swift.org/t/problem-with-clean-xcode-builds-on-new-stable-branches/29498/7)

But at this point, you mention it still didn't record a force downcast fix, the only fix I could see is the `as to as!` which is a `CoerceToCheckedCast` recorded as a disjunction choice in `addExplicitConversionConstraint`. Inside this, it fails because it defers a diagnostic for Unresolved checked cast should be handled for `diagnoseFailureForExpr` which falls down to CSDiag...

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Also, I confirmed that the diagnostics are actually being emitted via a fix added in `repairFailures`. But only when doing the re-type check. What is happening is that on the first type checking [here](https://github.com/apple/swift/blob/b32e82c720f6b6a5150fb09c63292f6709947d11/lib/Sema/TypeCheckStmt.cpp#L1794) there's no convertTo type being provided, so it skips a conversion constraint addition [here](https://github.com/apple/swift/blob/master/lib/Sema/CSSolver.cpp#L1232), therefore not getting into the `repairFailures` so the only fix attempted is the `[fix: as to as!]` I mentioned before. So when re-type checking it is able to pass the convertTo type [here](https://github.com/apple/swift/blob/master/lib/Sema/CSDiag.cpp#L3550), making it adds the conversion constraint and getting into `repairFailures` where the corrected fixes are already being recorded and correctly emitted. But all that after re-type check :(( 

And again @xedin thank's for all the reviews and for all the patience in those PR's, since I'm learning those things, I may get some things wrong sometimes :)

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: I'll resume working on this as soon as I can get the build right 👍

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: You are right, the actual fix is `CoerceToCheckedCast`. I think the reason why it's not diagnosed is due to `typeCheckCheckedCast` returning `CheckedCastKind::Unresolved`. As I mentioned before I think we kind of stumbled into a bigger problem here - `as to as!` _shouldn't_ be recorded if it doesn't help. It might be a good idea to not record a `CoerceToCheckedCast` as a disjunction choice at all, but attempt it if `conversion` choice fails and use `typeCheckCheckedCast` to verify that types are indeed related. And thanks for helping with this, I really appreciate it!

**Comment by LucianoPAlmeida** [comment]
@xedin Thanks to the guys on the forums who have been working on the [build issues](https://forums.swift.org/t/problem-with-clean-xcode-builds-on-new-stable-branches/29498/25) I got back to work on this today \o/
Few points:
Just removing the `CoerceToCheckedCast` from the disjunction and attempting that later lead to no only fix the issue, but also to an improvement in the diagnostics for some cases (specially coercion involving generic types or protocols).
Can you take a look when you have some time? :)

**Comment by xedin** [comment]
@LucianoPAlmeida Sounds great! This is in my queue to take a look very soon.

File: lib/Sema/CSDiagnostics.cpp
Comment (xedin) [inline lib/Sema/CSDiagnostics.cpp]: Let's just `return false;` in this case.

**Review by xedin — APPROVED** [review]
Awesome! I have left a tiny comment and afterwards I think this is ready to go! @LucianoPAlmeida would you mind searching for JIRA bugs with could be addressing? I'm pretty sure there are a few :)

File: lib/Sema/CSDiagnostics.cpp
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: Sure, Done :)

**Comment by LucianoPAlmeida** [comment]
@xedin Thank you for the review, I'll search for JIRA issues right now :)) 
They are probably marked as DiagnosticsQol label so I can filter for that?

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@LucianoPAlmeida I was hoping that there are going to be some which mention that `force downcast` suggestion is incorrect, but thank you for trying anyway! :)

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**
